### PR TITLE
[llvm-readobj] Print AMDGPU specific values in GNU mode

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
Currently, we don't emit any of the AMDGPU specific flags and options
when printing in GNU-mode (the default). This patch adds extra code to
handle outputting these values to match the output from GNU's `readelf`
when used on an AMDGPU image.

### CODE REVIEW

File: llvm/tools/llvm-readobj/ELFDumper.cpp [hunk 1]
```diff
@@ -3630,6 +3641,28 @@ template <class ELFT> void GNUELFDumper<ELFT>::printFileHeaders() {
...
+    switch (e.e_ident[ELF::EI_ABIVERSION]) {
+    default:
+      break;
+    case 0:
+      // ELFOSABI_AMDGPU_PAL, ELFOSABI_AMDGPU_MESA3D support *_V3 flags.
+      [[fallthrough]];
+    case ELF::ELFABIVERSION_AMDGPU_HSA_V3:
+      ElfFlags =
+          printFlags(e.e_flags, ArrayRef(ElfHeaderAMDGPUFlagsABIVersion3),
+                     unsigned(ELF::EF_AMDGPU_MACH));
```
Comment (arsenm) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:3654]: weird constructor cast style

File: llvm/tools/llvm-readobj/ELFDumper.cpp [hunk 2]
```diff
@@ -3630,6 +3641,28 @@ template <class ELFT> void GNUELFDumper<ELFT>::printFileHeaders() {
...
+  else if (e.e_machine == EM_AMDGPU) {
+    switch (e.e_ident[ELF::EI_ABIVERSION]) {
+    default:
+      break;
+    case 0:
+      // ELFOSABI_AMDGPU_PAL, ELFOSABI_AMDGPU_MESA3D support *_V3 flags.
+      [[fallthrough]];
+    case ELF::ELFABIVERSION_AMDGPU_HSA_V3:
+      ElfFlags =
+          printFlags(e.e_flags, ArrayRef(ElfHeaderAMDGPUFlagsABIVersion3),
```
Comment (arsenm) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:3653]: I'm surprised you need the ArrayRef

File: llvm/test/tools/llvm-readobj/ELF/file-header-os-abi.test [hunk 3]
```diff
@@ -167,7 +167,7 @@ FileHeader:
 # RUN: llvm-readelf --file-headers %t.osabi.amd.hsa | FileCheck %s --match-full-lines --check-prefix=OSABI-AMDHSA-GNU
 
 # OSABI-AMDHSA-LLVM: OS/ABI: AMDGPU_HSA (0x40)
-# OSABI-AMDHSA-GNU:  OS/ABI: 40
```
Comment (arsenm) [inline llvm/test/tools/llvm-readobj/ELF/file-header-os-abi.test:170]: Is there a bigger test diff?

Comment (jhuber6) [reply llvm/test/tools/llvm-readobj/ELF/file-header-os-abi.test:170]: Nope, only one that changed. I could make some new ones but it doesn't seem like this is tested strenuously anywhere else so I figured it was considered fairly trivial.

File: llvm/tools/llvm-readobj/ELFDumper.cpp (hunk 1 above)
Comment (jhuber6) [reply llvm/tools/llvm-readobj/ELFDumper.cpp:3654]: It's consistent with the other uses next to it, didn't want to rework anything.

File: llvm/tools/llvm-readobj/ELFDumper.cpp (hunk 2 above)
Comment (jhuber6) [reply llvm/tools/llvm-readobj/ELFDumper.cpp:3653]: Same as above, you don't need the `ArrayRef` to be explicitly cast, but the rest of them look like this so I figured I'd be consistent.

**Review by arsenm — APPROVED** [review]
Seems like there are missing tests

**Pull request merged**